### PR TITLE
Use AsyncLocal for ContextAppender

### PR DESCRIPTION
**Pull request by danielmarbach**
CallContext flows with the execution context. Meaning even if a Transport internally uses Task.Run the call context set from the acceptance test will float into it.

https://blogs.msdn.microsoft.com/pfxteam/2012/06/15/executioncontext-vs-synchronizationcontext/

Arguably it might be better to use AsyncLocal and upgrade to 4.6 and higher since CallContext might be not Core compliant (haven't checked).  I would prefer the AsyncLocal solution.

With that, we can make sure there is always a ScenarioContext available without being exposed to the inlining of the static context assignment into the factory and then run into races where scenario context is suddenly null.

### CODE REVIEW

**Comment by andreasohlund** [comment]
the 4.6 requirement would only be for the acceptance test fx right? (Should not be an issue to require 4.6 while keeping core at 4.5.2 should we want that?)

**Comment by danielmarbach** [comment]
@andreasohlund why would we constrain ourselves to 4.5.2? We are talking about develop which is next major right?

**Comment by andreasohlund** [comment]
Not against bumping to 4.6

**Comment by danielmarbach** [comment]
Saw the argument around 4.6.x vs 4.5.2. Happy with that decision. So then let's stay with CallContext if it receives @bording @timbussmann blessing in terms of compatible with core.

**Comment by danielmarbach** [comment]
Ping @Scooletz

**Comment by bording** [comment]
None of the `System.Runtime.Remoting` namespace is available in .NET Core, so this isn't going to work.

**Comment by danielmarbach** [comment]
Well guess then we have to think whether:

- We want to bump acceptance test FW dependency only and inflict pain on us only and live with mixed FW inside same solution or
- Bump everything

**Comment by Scooletz** [comment]
I think there's a third option @danielmarbach We could use `#if` to determine whether we should use `CallContext`/`ExecutionContext` or `AsyncLocal` for .NET Standard 2.0, right?

**Comment by bording** [comment]
We could certainly do some conditional compilation if we have to. In the main NSB codebase, I want to avoid that as much as possible, but for the testing framework that would probably be OK.

**Comment by Scooletz** [comment]
This one condition would enable us to:
- don't bump up the FW version for none of the components
- handle `static`s properly (and moreover, prepare for parallel execution of tests)
- maintain Standard preparedness

**Comment by timbussmann** [comment]
sounds like bumping the acceptance testing FW would be the easiest option?

**Comment by bording** [comment]
I don't care for the "bump just the testing framework to 4.6" approach because that means that all testing projects will also have to be built against 4.6, so we won't actually be able to run tests against the actual minimum version NServiceBus supports, 4.5.2.

**Comment by bording** [comment]
So, despite my previous comment, after talking with the other @Particular/nservicebus-maintainers today in our sync meeting, we came to the conclusion that being able to test on 4.5.2 specifically isn't worth the extra complexity of maintaining two separate implementations.

Since we're talking about bumping the version for the acceptance testing project anyway, we might as well bump up to 4.6.2 instead of just 4.6. 

So, the current plan is go ahead and bump it up to 4.6.2 and use AsyncLocal to implement this. @danielmarbach are you interested in making these changes?

**Comment by Scooletz** [comment]
Whoever will be implementing it, please ping me for a review. I'd like to take a look at all this `async` goodness!

**Comment by danielmarbach** [comment]
@Scooletz if you have capacity feel free to move ahead. I'm pretty swamped with a few other things so it will take at least 2 weeks until I can do this.

**Comment by timbussmann** [comment]
@Scooletz in case you have no capacity as well, please let us know so we can try to find someone so that this PR can be moved forward timely.

**Comment by Scooletz** [comment]
Ping @Particular/nservicebus-maintainers 

I let myself try the `polyfill` approach proposed by me. I find it simple and elegant. It does not require bumping up and with conditional compilation `#if NET452` is `Standard-ready`.

**Comment by bording** [comment]
@Scooletz I don't think any of us realized how straightforward the polyfill approach would be. 👍 for doing this over bumping the framework version.

File: src/NServiceBus.AcceptanceTesting/ScenarioContext.cs [hunk 1]
```diff
@@ -46,5 +47,17 @@ public override string ToString()
                 return $"{Level}: {Message}";
             }
         }
+
+        static readonly AsyncLocal<ScenarioContext> asyncContext = new AsyncLocal<ScenarioContext>();
```
Comment (bording) [inline src/NServiceBus.AcceptanceTesting/ScenarioContext.cs]: From a formatting standpoint, this declaration should probably be above the `LogItem` declaration.

File: src/NServiceBus.AcceptanceTesting/ScenarioContext.cs [hunk 2]
```diff
@@ -46,5 +47,17 @@ public override string ToString()
...
+
+        internal static ScenarioContext GetContext()
```
Comment (bording) [inline src/NServiceBus.AcceptanceTesting/ScenarioContext.cs]: Can we make these a static `Current` property instead?

That would also mean this should be moved up to the top of the file.

File: src/NServiceBus.AcceptanceTesting/ContextAppender.cs [hunk 3]
```diff
@@ -7,17 +7,11 @@
     // This class is written under the assumption that acceptance tests are executed sequentially.
```
Comment (bording) [inline src/NServiceBus.AcceptanceTesting/ContextAppender.cs]: I believe this comment is no longer true with these changes, correct? Should be able to remove this.

File: src/NServiceBus.AcceptanceTesting/AsyncLocal.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
+﻿#if NET452
+namespace System.Threading
+{
+    using Runtime.Remoting.Messaging;
+
+    /// <summary>
+    /// Provides a polyfill of AsyncLocal in .NET 4.5.2
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
```
Comment (Scooletz) [inline src/NServiceBus.AcceptanceTesting/AsyncLocal.cs]: Will do.

File: src/NServiceBus.AcceptanceTesting/ScenarioContext.cs (hunk 1 above)
Comment (Scooletz) [inline src/NServiceBus.AcceptanceTesting/ScenarioContext.cs]: Will reformat.

File: src/NServiceBus.AcceptanceTesting/ScenarioContext.cs (hunk 2 above)
Comment (Scooletz) [inline src/NServiceBus.AcceptanceTesting/ScenarioContext.cs]: Sure. Will refactor.

File: src/NServiceBus.AcceptanceTesting/ContextAppender.cs (hunk 3 above)
Comment (Scooletz) [inline src/NServiceBus.AcceptanceTesting/ContextAppender.cs]: Correct. Now with the async-aware sticky context every task that is spawned will have (as its descendants) properly assigned `ScenarioContext`

Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTesting/ContextAppender.cs]: Yeah this comment can be removed

**Comment by Scooletz** [comment]
Ping @bording 
I think I addressed all the issues.

File: src/NServiceBus.AcceptanceTesting/AsyncLocal.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+namespace System.Threading
+{
+    using Runtime.Remoting.Messaging;
+
+    // Provides a polyfill of AsyncLocal in .NET 4.5.2
+    sealed class AsyncLocal<T>
+    {
+        readonly string id;
+
+        /// <summary>Gets or sets the value of the ambient data. </summary>
```
Comment (bording) [inline src/NServiceBus.AcceptanceTesting/AsyncLocal.cs]: Some more xml comments to remove!

File: src/NServiceBus.AcceptanceTesting/AsyncLocal.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        /// <returns>The value of the ambient data. </returns>
+        public T Value
+        {
+            get
+            {
+                var localValue = CallContext.LogicalGetData(id);
+                if (localValue != null)
+                    return (T)localValue;
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTesting/AsyncLocal.cs]: needs some more brackets.

**Comment by Scooletz** [comment]
@timbussmann @bording 
Changes applied.

**Pull request merged**